FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/DraftViewModel.kt

```diff
@@ -0,0 +1,301 @@
+            KeyStoreCryptoManager.encryptOutputStream(outputStream) { cipherOutputStream ->
+              mimeMessage.writeTo(cipherOutputStream)
+            }
+          }
+
+          MsgsCacheManager.storeMsg(draftMessageEntity.id.toString(), mimeMessage as MimeMessage)
+          val messageEntityWithoutStateChange = draftMessageEntity.copy(
+            subject = outgoingMessageInfo.subject,
+            fromAddress = InternetAddress.toString(arrayOf(outgoingMessageInfo.from)),
+            replyTo = InternetAddress.toString(arrayOf(outgoingMessageInfo.from)),
+            toAddress = InternetAddress.toString(outgoingMessageInfo.toRecipients?.toTypedArray()),
+            ccAddress = InternetAddress.toString(outgoingMessageInfo.ccRecipients?.toTypedArray()),
+            sentDate = mimeMessage.sentDate?.time,
+            receivedDate = mimeMessage.sentDate?.time
+          )
+
+          roomDatabase.msgDao().updateSuspend(
+            if (isDeleted) {
+              messageEntityWithoutStateChange
+            } else {
+              messageEntityWithoutStateChange.copy(state = MessageState.PENDING_UPLOADING_DRAFT.value)
+            }
+          )
+          return@withContext Result.success(true)
+        } ?: return@withContext Result.success(false)
+      } catch (e: Exception) {
+        e.printStackTrace()
+        return@withContext Result.exception(e)
+      }
+    }
+
+  private suspend fun genDraftMessageEntity(
+    accountEntity: AccountEntity,
+    outgoingMessageInfo: OutgoingMessageInfo
+  ): MessageEntity =
+    withContext(Dispatchers.IO) {
+      val foldersManager = FoldersManager.fromDatabaseSuspend(getApplication(), accountEntity)
+      val folderDrafts =
+        foldersManager.folderDrafts ?: throw IllegalStateException("Drafts folder is undefined")
+      val newDraftMessageEntity = MessageEntity.genMsgEntity(
+        email = accountEntity.email,
+        label = folderDrafts.fullName,
+        uid = System.currentTimeMillis(),
+        info = outgoingMessageInfo,
+        flags = listOf(MessageFlag.DRAFT, MessageFlag.SEEN)
+      ).copy(
+        state = MessageState.PENDING_UPLOADING_DRAFT.value,
+        threadId = gmailThreadId
+      )
+      val id = roomDatabase.msgDao().insertSuspend(newDraftMessageEntity)
+      return@withContext newDraftMessageEntity.copy(
+        id = id,
+        receivedDate = newDraftMessageEntity.sentDate
+      )
+    }
+
+  fun deleteDraft(coroutineScope: CoroutineScope = viewModelScope) {
+    isDeleted = true
+    coroutineScope.launch {
+      sessionDraftMessageEntity?.let { draftEntity ->
+        roomDatabase.msgDao().updateSuspend(
+          draftEntity.copy(state = MessageState.PENDING_DELETING_DRAFT.value)
+        )
+
+        DeleteDraftsWorker.enqueue(getApplication())
+      }
+    }
+  }
+
+  private data class DraftFingerprint(
+    var msgText: String? = null,
+    var msgSubject: String? = null,
+    val toRecipients: Set<String> = mutableSetOf(),
```

Comment:
why not just `= setOf<String>()`

Reply:
![image](https://user-images.githubusercontent.com/2863246/193994337-2116f676-6adb-4b4f-8445-3cd0d0fdce55.png)